File: src/ca/mindmagic/game/map/grid/TiledGridMap.java
package ca.mindmagic.game.map.grid;

import ca.mindmagic.game.map.grid.pattern.Pattern;
import kotlin.collections.SetsKt;

import java.util.*;
import java.util.function.Supplier;

public class TiledGridMap<T> extends GridMap {

    ArrayList<T> tiles;
    Map<T, Set<T>> neighborMap = new HashMap<>();

    public TiledGridMap(Pattern pattern, int width, int height, boolean wrapHorizontal, boolean wrapVertical, Supplier<T> tileType) {
        super(pattern, height, width, wrapHorizontal, wrapVertical);
        tiles = new ArrayList<>(width * height);
        initializeTiles(tileType);
        mapNeighbors();
    }

    private void initializeTiles(Supplier<T> tileType) {
        for (int i = 0; i < super.getWidth() * super.getHeight(); i++) {
            tiles.add(tileType.get());
        }
    }

    private void mapNeighbors() {
        tiles.forEach(t -> neighborMap.put(t, calculateNeighbors(t)));
    }

    private Set<T> calculateNeighbors(T tile) {
        return SetsKt.emptySet();
    }

    public T getTile(int row, int col) {
        if (row > super.getHeight() || col > super.getWidth()) return null;
        return tiles.get(row * super.getWidth() + col);
    }

    public Set<T> getNeighborsOf(T tile) {
        return neighborMap.get(tile);
    }
}

File: src/ca/mindmagic/game/map/grid/TiledGridMap.kt
package ca.mindmagic.game.map.grid

import ca.mindmagic.game.map.grid.pattern.Pattern

import java.util.*
import java.util.function.Supplier
import java.util.stream.Collectors

class TiledGridMap<T>(pattern: Pattern, width: Int, height: Int, wrapHorizontal: Boolean, wrapVertical: Boolean, tileType: Supplier<T>) : GridMap(pattern, height, width, wrapHorizontal, wrapVertical) {

    private val tiles = ArrayList<T>(width * height)
    private val neighborMap: MutableMap<T, Set<T>> = HashMap()

    init {
        initializeTiles(tileType)
        mapNeighbors()
    }

    private fun initializeTiles(tileType: Supplier<T>) {
        for (i in 0 until super.width * super.height) {
            tiles.add(tileType.get())
        }
    }

    private fun mapNeighbors() {
        tiles.forEach { t -> neighborMap[t] = calculateNeighbors(t) }
    }

    private fun calculateNeighbors(tile: T): Set<T> {
        val neighbors = neighborsOf(coordinateOf(tile))
        return neighbors.stream()
                .map { c -> tileAt(c) }
                .collect(Collectors.toSet())
    }

    fun tileAt(location: Coordinate) = tiles[location.row * super.width + location.col]

    fun neighborsOf(tile: T): Set<T> {
        return neighborMap[tile] ?: emptySet()
    }

    fun range(source: T, target: T): Int {
        return range(coordinateOf(source), coordinateOf(target))
    }

    fun coordinateOf(tile: T): Coordinate {
        val index = tiles.indexOf(tile);
        return Coordinate(row(index), col(index))
    }

    private fun row(index: Int) = index / width

    private fun col(index: Int) = index % width


}
